### DISCUSSION TITLE
[Urgent] Session recordings takes up too much storage space

### DISCUSSION THREAD

**Discussion by kangeek** [gravitational/teleport#26904] [Q&A] [answered]
### Summary

We deployed teleport cluster on Azure AKS, with shared volume mounted to all of teleport-cluster-auth pods on `/var/lib/teleport`.

We gave 500G of space to that volume, and it is 100% used !!! We would like to know if it is normal. These are some information of our deployment:

- Teleport cluster and agent of version `12.4.3`.
- Teleport cluster deployed on K8S, with 3 replicas of auth service and proxy service, cluster state stored in a 3 replica etcd cluster.
- Teleport agents: ssh-agents on 150 VMs, teleport-kube-agent on 15 K8S clusters.
- Session recording is enabled.
- 24 RBAC roles with `spec.options.enhanced_recording: [command, network]`.
- `/var/lib/teleport/log/records/multi/` of teleport-cluster-auth takes up almost all of the space. 

My question is: **considering our usage scenario, is it normal that session recordings take up all of the 500G storage, and how can I reclaim ocupied space?.**

### Details

Nowadays, we found that a lot of VM agents keeps reporting ERRORs:

```
2023-05-25T23:01:08+08:00 WARN [UPLOAD]    Backing off, will retry after 10s. error:[
ERROR REPORT:
Original Error: *trace.ConnectionProblemError timeout waiting for stream status update
Stack Trace:
        github.com/gravitational/teleport/lib/events/filesessions/fileasync.go:496 github.com/gravitational/teleport/lib/events/filesessions.(*Uploader).upload
        github.com/gravitational/teleport/lib/events/filesessions/fileasync.go:427 github.com/gravitational/teleport/lib/events/filesessions.(*Uploader).startUpload.func1
        runtime/asm_amd64.s:1598 runtime.goexit
User Message: timeout waiting for stream status update] filesessions/fileasync.go:192
```

And we found that the teleport cluster not running smoothly as usual. Finally, we found that it was because the shared volume mounted on `/var/lib/teleport` of teleport-cluster-auth pods is full. Some phenomena:

- Agents with the error log as above.
- when we check the "Session Recordings" and "Audit Log" under "Management" on WebUI, they are empty. That means the recordings and logs are not uploaded. **But we checked the day 7 days ago, the recordings and logs exist.**
- Then we checked the teleport cluster pods, teleport-cluster-auth restarted several times with this event: 

```
Events:
  Type     Reason              Age                    From     Message
  ----     ------              ----                   ----     -------
  Warning  VolumeResizeFailed  2m47s (x362 over 12h)  kubelet  NodeExpandVolume.NodeExpandVolume failed for volume "pvc-974d1027-4419-4479-af4a-b9423badf873" : Expander.NodeExpand found CSI plugin kubernetes.io/csi/file.csi.azure.com to not support node expansion
```

> `file.csi.azure.com` is the Azure provisioner with Azure File Storage as backend.

At that moment, the "500G shared volume" I mentioned earlier was 100G. So I expanded the volume to 200G, and it was full soon after; then I expanded to 300G ... 400G ... 500G, and it was full aggin and again. **The whole expansion process took place in just three to four hours.

And we stopped here, this is the situation now, we don't know how much more space it will cost.

**We don't know wheather the volume had been full a long time ago, or just recent days (since the day 7 days ago, the recordings and logs exist). If the latter, why it take only a few hours to consume all of the expanded space?**

Under `/var/lib/teleport/logs/records`, there are 23000+ `*.tar` files, along with a folder `multi`. Under `/var/lib/teleport/log/records/multi/`, there are 110000+ folders and each folder takes up 5Mi space. That is why it consumes soooo much space!

**What does this `/var/lib/teleport/log/records/multi/` folder contains? How can I reclaim occupied disk space? Can I just delete this whold folder?**

### Thanks

The errors is still going on, so this is URGENT. I look forward to and appreciate any help.

**Comment by webvictim** [comment, marked as answer]
This is not normal at all. You say that you're using a volume for `/var/lib/teleport` which is shared between multiple pods? This is likely the reason for your issues - the use of shared `PersistentVolumeClaims` or similar for audit log/session recording storage is not supported and will result in errors.

`etcd` only stores the cluster's backend state (users, roles etc) in an HA fashion - in this scenario, audit logs will still be written to `/var/lib/teleport/log` and session recordings will go to `/var/lib/teleport/log/records`.  If the volume with these directories on is shared between multiple pods, they are probably all contending with each other's data and overriding it, resulting in these issues.

This warning about not configuring claims like this is noted in the values file here: https://github.com/gravitational/teleport/blob/6bfa871ddeb1c0543ac5d5e0cdf3c67f5d5e5ca0/examples/chart/teleport-cluster/values.yaml#L376-L377

It's also in the Helm reference: https://goteleport.com/docs/reference/helm-reference/teleport-cluster/#for-auth-pods

Your best bet here is likely to:
- scale the Teleport cluster down to 0 replicas temporarily
- remove the contents of `/var/lib/teleport/log/records/multi` on the `PersistentVolumeClaim` (note: you will likely lose some recordings for sessions that took place during these periods of high activity, but that's unfortunately just how it is)
- scale the cluster back up to just 1 auth server replica (you can still have multiple proxy replicas)
- test logins and make sure that things work

Your nodes should theoretically be able to upload any session recordings they have stored locally to the auth server. Things will be a lot slower with just one pod and you will probably see high CPU usage, but as long as your worker nodes are reasonably powerful it should be able to handle 150 nodes and a handful of concurrent sessions.

Longer term, you should probably look to move your audit log and session recording storage to a better supported HA solution like AWS DynamoDB or Google Firestore. Failing that, you could fork the Helm chart to use a `VolumeClaimTemplate` to mount individual storage to each auth pod similar to what this PR is doing: https://github.com/gravitational/teleport/pull/24299/

We don't have a fully native Azure solution for the Teleport backend and audit log yet, but we're working on it and should have something out within a few months. Maybe one of these workarounds could get you through until that's released.

**Reply by kangeek** [reply]
Thanks for your patient and comprehensive reply. I will try your method asap.

**Reply by kangeek** [reply]
Sorry to bother again, I did as you said:

> - scale the Teleport cluster down to 0 replicas temporarily
> - remove the contents of `/var/lib/teleport/log/records/multi` on the PersistentVolumeClaim 
> - scale the cluster back up to just 1 auth server replica 
> - test logins and make sure that things work, and it works

And the only one auth pod still generates files under `/var/lib/teleport/log/records/multi` like mad, the 500G disk is nearly full now.

The auth pod keeps restarting:

```
Events:
  Type     Reason     Age                      From     Message
  ----     ------     ----                     ----     -------
  Warning  Unhealthy  73m (x25 over 7h5m)      kubelet  (combined from similar events): Readiness probe failed: Get "http://192.168.80.11:3000/readyz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
  Warning  Unhealthy  14m (x240 over 7h57m)    kubelet  Liveness probe failed: Get "http://192.168.80.11:3000/healthz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
  Warning  Unhealthy  5m15s (x247 over 7h57m)  kubelet  Readiness probe failed: Get "http://192.168.80.11:3000/readyz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
  Warning  BackOff    9s (x1476 over 7h55m)    kubelet  Back-off restarting failed container
```

The only one auth pod keeps output stack trace like this:

```
goroutine 23294 [runnable]:
syscall.Syscall6(0xc0156791d0?, 0x23?, 0x7efff8411018?, 0x7f0031a24108?, 0xd0?, 0xc012ab2c00?, 0xc0139a0b60?)
	syscall/syscall_linux.go:91 +0x36 fp=0xc03786b958 sp=0xc03786b8d0 pc=0x5f04b6
syscall.fstatat(0x23?, {0xc0156791d0?, 0x0?}, 0xc0139a0b98, 0xc0156791d0?)
	syscall/zsyscall_linux_amd64.go:1443 +0x9a fp=0xc03786b9d0 sp=0xc03786b958 pc=0x5ef6da
syscall.Stat(...)
	syscall/syscall_linux_amd64.go:66
os.statNolog.func1(...)
	os/stat_unix.go:32
os.ignoringEINTR(...)
	os/file_posix.go:245
os.statNolog({0xc0156791d0, 0x23})
	os/stat_unix.go:31 +0x58 fp=0xc03786ba20 sp=0xc03786b9d0 pc=0x629b98
os.Stat({0xc0156791d0, 0x23})
	os/stat.go:13 +0x34 fp=0xc03786ba40 sp=0xc03786ba20 pc=0x6296d4
os.MkdirAll({0xc0156791d0, 0x23}, 0x3b437be0?)
	os/path.go:20 +0x2b fp=0xc03786ba80 sp=0xc03786ba40 pc=0x627e6b
github.com/gravitational/teleport/lib/events/filesessions.(*Handler).CreateUpload(0xc0010b0ce0, {0x7b81c20?, 0xc0010e19d0?}, {0xc015678f90, 0x24})
	github.com/gravitational/teleport/lib/events/filesessions/filestream.go:84 +0xa5 fp=0xc03786bb58 sp=0xc03786ba80 pc=0x3b1b4c5
github.com/gravitational/teleport/lib/events.(*ProtoStreamer).CreateAuditStream(0xc0010f46f0, {0x7b783e8, 0xc02ebf4990}, {0xc015678f90, 0x24})
	github.com/gravitational/teleport/lib/events/stream.go:148 +0x5e fp=0xc03786bc08 sp=0xc03786bb58 pc=0x2a8aefe
github.com/gravitational/teleport/lib/events.(*CheckingStreamer).CreateAuditStream(0xc0010f2100, {0x7b783e8?, 0xc02ebf4990?}, {0xc015678f90?, 0x1?})
	github.com/gravitational/teleport/lib/events/emitter.go:341 +0x3c fp=0xc03786bc50 sp=0xc03786bc08 pc=0x2a80cbc
github.com/gravitational/teleport/lib/events.(*ReportingStreamer).CreateAuditStream(0xc0010fa060, {0x7b783e8?, 0xc02ebf4990?}, {0xc015678f90, 0x24})
	github.com/gravitational/teleport/lib/events/emitter.go:644 +0x42 fp=0xc03786bc98 sp=0xc03786bc50 pc=0x2a81de2
github.com/gravitational/teleport/lib/auth.(*Server).CreateAuditStream(0x5dac820?, {0x7b783e8, 0xc02ebf4990}, {0xc015678f90, 0x24})
	github.com/gravitational/teleport/lib/auth/auth.go:3964 +0x92 fp=0xc03786bcd8 sp=0xc03786bc98 pc=0x38371d2
github.com/gravitational/teleport/lib/auth.(*ServerWithRoles).CreateAuditStream(0xc00ce4f200, {0x7b783e8, 0xc02ebf4990}, {0xc015678f90, 0x24})
	github.com/gravitational/teleport/lib/auth/auth_with_roles.go:3364 +0x250 fp=0xc03786caf8 sp=0xc03786bcd8 pc=0x3862830
github.com/gravitational/teleport/lib/auth.(*GRPCServer).CreateAuditStream(0xc000edc360, {0x7b88000, 0xc03b4368a0})
	github.com/gravitational/teleport/lib/auth/grpcserver.go:237 +0x3a7 fp=0xc03786d310 sp=0xc03786caf8 pc=0x388d147
github.com/gravitational/teleport/api/client/proto._AuthService_CreateAuditStream_Handler({0x68fde00?, 0xc000edc360}, {0x7b81878?, 0xc042a3e940})
	github.com/gravitational/teleport/api@v0.0.0/client/proto/authservice.pb.go:20326 +0xa2 fp=0xc03786d348 sp=0xc03786d310 pc=0x122c122
github.com/gravitational/teleport/lib/auth.(*Middleware).withAuthenticatedUserStreamInterceptor(0xc0065a33b0?, {0x68fde00, 0xc000edc360}, {0x7b82370, 0xc02ebf4720}, 0x5db9780?, 0x72a7210)
	github.com/gravitational/teleport/lib/auth/middleware.go:426 +0xe7 fp=0xc03786d398 sp=0xc03786d348 pc=0x38d5447
github.com/gravitational/teleport/lib/auth.(*Middleware).withAuthenticatedUserStreamInterceptor-fm({0x68fde00?, 0xc000edc360?}, {0x7b82370?, 0xc02ebf4720?}, 0xc000fce070?, 0xc0065a33f8?)
	<autogenerated>:1 +0x5b fp=0xc03786d3e0 sp=0xc03786d398 pc=0x39beb7b
github.com/gravitational/teleport/lib/utils.ChainStreamServerInterceptors.func1.1({0x68fde00?, 0xc000edc360?}, {0xd?, 0x1?})
	github.com/gravitational/teleport/lib/utils/grpc.go:125 +0x38 fp=0xc03786d420 sp=0xc03786d3e0 pc=0x1659158
github.com/gravitational/teleport/lib/limiter.(*Limiter).StreamServerInterceptor(0xc000fc4560, {0x68fde00, 0xc000edc360}, {0x7b82370, 0xc02ebf4720?}, 0x7f0031a24108?, 0xc02ebf4750)
	github.com/gravitational/teleport/lib/limiter/limiter.go:176 +0x1ea fp=0xc03786d4b0 sp=0xc03786d420 pc=0x16bbb2a
github.com/gravitational/teleport/lib/limiter.(*Limiter).StreamServerInterceptor-fm({0x68fde00?, 0xc000edc360?}, {0x7b82370?, 0xc02ebf4720?}, 0x635dec0?, 0xc02ebf4801?)
	<autogenerated>:1 +0x5b fp=0xc03786d4f8 sp=0xc03786d4b0 pc=0x39bea9b
github.com/gravitational/teleport/lib/utils.ChainStreamServerInterceptors.func1.1({0x68fde00?, 0xc000edc360?}, {0x7effd40531a8?, 0x7f0031a24108?})
	github.com/gravitational/teleport/lib/utils/grpc.go:125 +0x38 fp=0xc03786d538 sp=0xc03786d4f8 pc=0x1659158
github.com/gravitational/teleport/lib/utils.GRPCServerStreamErrorInterceptor({0x68fde00, 0xc000edc360}, {0x7b82370?, 0xc02ebf4720}, 0x64e6b40?, 0xc02ebf4780)
	github.com/gravitational/teleport/lib/utils/grpc.go:78 +0x8e fp=0xc03786d570 sp=0xc03786d538 pc=0x1658a0e
github.com/gravitational/teleport/lib/utils.ChainStreamServerInterceptors.func1.1({0x68fde00?, 0xc000edc360?}, {0x696a53d?, 0xb?})
	github.com/gravitational/teleport/lib/utils/grpc.go:125 +0x38 fp=0xc03786d5b0 sp=0xc03786d570 pc=0x1659158
github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors.StreamServerInterceptor.func1({0x68fde00, 0xc000edc360}, {0x7b82370, 0xc02ebf4720}, 0xc01703a108, 0xc02ebf47b0)
	github.com/grpc-ecosystem/go-grpc-middleware/v2@v2.0.0-rc.2.0.20220308023801-e4a6915ea237/interceptors/server.go:35 +0x31e fp=0xc03786d720 sp=0xc03786d5b0 pc=0xaf253e
github.com/gravitational/teleport/lib/utils.ChainStreamServerInterceptors.func1({0x68fde00, 0xc000edc360}, {0x7b82370?, 0xc02ebf4720?}, 0xc01703a108, 0x0?)
	github.com/gravitational/teleport/lib/utils/grpc.go:129 +0x15b fp=0xc03786d798 sp=0xc03786d720 pc=0x16590bb
google.golang.org/grpc.getChainStreamHandler.func1({0x68fde00, 0xc000edc360}, {0x7b82370, 0xc02ebf4720})
	google.golang.org/grpc@v1.52.0/server.go:1478 +0xb9 fp=0xc03786d800 sp=0xc03786d798 pc=0xadedd9
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.StreamServerInterceptor.func1({0x68fde00, 0xc000edc360}, {0x7b82520, 0xc01f07d0a0}, 0xc01703a108, 0xc01e727500)
	go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.38.0/interceptor.go:435 +0x43a fp=0xc03786dae0 sp=0xc03786d800 pc=0x2aa727a
google.golang.org/grpc.chainStreamInterceptors.func1({0x68fde00, 0xc000edc360}, {0x7b82520, 0xc01f07d0a0}, 0x5940c80?, 0xc03b436600?)
	google.golang.org/grpc@v1.52.0/server.go:1469 +0x8f fp=0xc03786db38 sp=0xc03786dae0 pc=0xadeb6f
google.golang.org/grpc.(*Server).processStreamingRPC(0xc00079d680, {0x7b8b540, 0xc03da18b60}, 0xc02c5e7b00, 0xc0015ba4b0, 0xae05160, 0x0)
	google.golang.org/grpc@v1.52.0/server.go:1627 +0x1226 fp=0xc03786de48 sp=0xc03786db38 pc=0xae0046
google.golang.org/grpc.(*Server).handleStream(0xc00079d680, {0x7b8b540, 0xc03da18b60}, 0xc02c5e7b00, 0x0)
	google.golang.org/grpc@v1.52.0/server.go:1708 +0x9f0 fp=0xc03786df68 sp=0xc03786de48 pc=0xae1930
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.52.0/server.go:965 +0x98 fp=0xc03786dfe0 sp=0xc03786df68 pc=0xada2b8
runtime.goexit()
	runtime/asm_amd64.s:1598 +0x1 fp=0xc03786dfe8 sp=0xc03786dfe0 pc=0x5d8441
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.52.0/server.go:963 +0x28a
```

Did I miss something ?

**Reply by webvictim** [reply]
My best guess is that one or more of your nodes is spam-uploading session recordings. I don't quite understand why that'd be, but if you can get a node ID from the logs you might be able to connect to the node directly and clear out the locally-stored recordings that it's trying to upload.

**Reply by kangeek** [reply]
Your guess is right, and this is very helpful !

I checked the "Audit log" of type "Session Uploaded", and found that the day before the "storage-space-occupied" issue came out, one of my collegue executed a set of [ansible playbooks through teleport](https://goteleport.com/docs/server-access/guides/ansible/), and that generates 3700+ of "non-interactive" session recordings:

<img width="940" alt="image" src="https://github.com/gravitational/teleport/assets/7991571/e8db1e84-3e7b-4844-b221-050f77867da4">

Then I did the following:

- Completely uninstalled teleport agent on the relative VMs that has "non-interactive sessions", removed all teleport related directories, including `/var/lib/teleport` (because I manage teleport agent via ansible, this is easier than "clearing out the locally-stored recordings")
- Scale the Teleport auth deployment down to 0 replicas temporarily
- Remove the contents of `/var/lib/teleport/log/records/multi` on the PersistentVolumeClaim
- Scale the auth deployment back up to just 1 auth server replica, and **this time the storage space remains calm**
- After confirming that the teleport cluster working well, scale the auth deployment up to 3, after a while, **the storage space did not fill up quickly again, and the auth pods does not restart constantly**

So it seems that **"executing ansible through teleport" is the real cause** ? 

And running teleport-auth deployment of multiple pods mounting the same shared StorageClass backended by Azure File seems OK (but I am not 100% sure), and explains why it has been deployed this way for more that 3 months, and the "storage space occpied" issue just came out recently after running smoothly all along.

**Reply by webvictim** [reply]
Thanks for following up and for providing the detail. I'm very glad everything is fixed for you now.

I'll try to follow up on this and figure out why all these short-running non-interactive sessions could DoS the cluster in the way they did.
